### PR TITLE
change the calculation of the nonlinear residuum so that fields with an ...

### PR DESCRIPTION
**Pull request by jdannberg**
...initial residuum of zero are not included

### CODE REVIEW

**Comment by tjhei** [comment]
Did you test that this works? If yes, can you turn that into a simple test/ please?

**Comment by jdannberg** [comment]
I tested it on my melt branch, and there it fixed the problem with the residuum. 
But it shouldn't be so difficult to set the temperature in on of our existing tests to zero, so sure, I can add a test.

**Comment by tjhei** [comment]
Great, the test would produce "residual -nan" without your fix. With this I am getting "residual 1" though. This is probably because the compositional residual is not decreasing (it is at 1e-10 to begin with, though).

So our way to compute the residual is flawed here. Asking for a reduction of the nonlinear residual in all components doesn't make any sense especially in a time dependent computation. @bangerth do you have a suggestion? 

@jdannberg can you please change "residuum" to "residual" in the filenames and add a changelog entry?

**Comment by jdannberg** [comment]
okay, changed.

**Comment by bangerth** [comment]
It's not quite clear to me how this can happen. is it, for example, because you have a constant temperature field (zero temperature residual) but a strain rate dependent viscosity (nonzero Stokes residual)?

I was thinking for a while that maybe we should just set the `initial_residual` to what we have in the second iteration then, but that doesn't help because in the case above we will always have a zero temperature residual. In this case, the right solution would be to exclude certain equations from the convergence determination altogether, as this patch does. Except that it doesn't work very well as @tjhei points out. @jdannberg: The residual in the last equation doesn't reduce because nothing is solved (it keeps doing zero iterations) -- can you figure out why?

But I think that @tjhei is right in principle: This way of computing convergence doesn't make much sense. The correct way would be to scale the residuals in some way and add them up, then verify convergence of this scaled sum. The problem, of course, is mostly with how to do the scaling. That might be more work.

**Comment by jdannberg** [comment]
Initially we had exactly the problem you described: a model setup that did not depend on the temperature, but that had a strain-rate dependent viscosity (and melt migration, so we needed the iterated IMPES anyway). We noticed that in such a case the residual will be nan, so we thought the field that had an initial residual of zero shouldn't be included in the calculation of the overall residual. 

This test does just that: it checks that the residual is not nan if the initial residual is zero. I just copied another testcase with iterated IMPES, and I haven't even checked what exactly the model setup is, because I thought it wasn't important and we already have another test that checks the residual calculated by the iterated IMPES solver (iterated_IMPES). 

But I would be happy to make a new testcase or improve my existing one once we have decided on how exactly we want to compute convergence or what else my testcase should test :-)

**Comment by tjhei** [comment]
Maybe we can use the norm of the right-hand sides to scale the residuals? This would correspond to the initial residual with a zero guess. This would ignore the fact that our current guess might be very good/perfect in some variable already.

**Comment by bangerth** [comment]
I've tried thinking about this, but I don't have a particular good idea either. I do think that Timo's suggestion of looking at
```
  sum_j   ||r_j||/||rhs_j||
```
(or the maximum of these quantities) is probably the right way to go. In a time stepping scheme, the right hand sides should never be zero (except if you have a zero field that stays zero, in which case one would probably have to exclude these equations). I don't know how easy or difficult it would be to implement something like this.

I'd be ok with the patch as is, given that it is progress in some direction at least. What do you think, @tjhei ?

**Comment by tjhei** [comment]
Juliane has been playing with this on our melt branch ([1], [2]). I think the first results are promising. We can ask her to extract it so we get this in mainline sooner and can see how it works if you like.

This also should work for stationary problems because if you start with u=0, it is identical to the current version. We needed to exclude the static pressure from the residual, but Juliane already implemented that.

[1]: https://github.com/jdannberg/aspect/commit/93313cf4b0b211813a3d08f6afd50bc30466264e 
[2]: https://github.com/jdannberg/aspect/commit/4cbb301d00bd03d0ca1830159198c06dadd42e74

**Comment by jdannberg** [comment]
Yes, sure, if you think the way we implemented this in our melt branch is a good solution, then I think it's a good idea to include this in mainline already, as it is not directly related to the melt transport. 

I would be happy extract this somehow, although I am not really sure how to do this with git, because I would have to exclude all the lines related to the melt transport from my commits.

**Comment by bangerth** [comment]
I don't think there's a good way of doing this in git other than copying by hand the relevant code. If the changes are scattered throughout a file or multiple functions, you can also create a diff file and simply delete those changes you don't want to apply. Then use the `patch` program to apply the changes to a checkout of mainline.

**Comment by jdannberg** [comment]
Okay, thanks! I ended up doing it by hand, but I tested that the residuals are the same as in our melt transport branch to make sure I really copied all of the relevant changes. 

I also didn't update the the test outputs so far, in case we still want to change something in the calculation. 

Another change that is also included is that the output of the residuals is now in the same order as the output above about the systems that are solved and the number of iterations. I found this the most intuitive way (as we don't even have output that tells us which residual belongs to which system), but it might be confusing for people who are already using the iterated IMPES solver, so maybe we should discuss if/how we want to include this change.

**Comment by tjhei** [comment]
Great, thank you! Can you please rebase to the current master, squash, and fix the test output (iterated_IMPES and iterated_IMPES_residual)?

I think we should also add a test for iterated_IMPES with the direct solver. Would you mind copying iterated_IMPES?

**Comment by gassmoeller** [comment]
I would argue for including your different ordering scheme of the residuals, because it is much more intuitive. Maybe you could include it in changes.h to keep people informed.

**Comment by jdannberg** [comment]
Okay, I added a new test, updated the output, rebased and squashed.

File: tests/iterated_IMPES_direct_solver.prm [hunk 1]
```diff
@@ -0,0 +1,141 @@
+#########################################################
+# This is a model setup to test the iterated IMPES
+# nonlinear solver scheme when using the direct solver. 
+# Both the calculation of the residuum and the convergence 
+# with prescribed boundary conditions are tested, as well 
+# as agreement with the respective test for the iterative
+# solver (iterted_IMPES).
```
Comment (tjhei) [inline tests/iterated_IMPES_direct_solver.prm]: typo: "iterated"

**Comment by tjhei** [comment]
Great job, thank you! Can you please fix the one typo?

**Comment by jdannberg** [comment]
Oh, somehow I missed that one, thank you!
It should be fixed now.

**Comment by bangerth** [comment]
Thanks, y'all for working on this!
